### ISSUE TITLE
Full File Diff

### ISSUE THREAD

**Issue by deckar01** [fedora-ruby/gem-compare#17]
## Problem

Since gems can be published with arbitrary content from the developer's machine, some projects require auditing the contents of the gems directly rather than the commits in the source repository. I currently fetch and unpack them, then run `git diff -U0` on them to get the file changes. `gem-compare` automates much of that process, but does not provide detailed insight into code content that changed.

```sh
$ gem compare --files deckar01-task_list 2.0.0 2.0.1
Fetching: deckar01-task_list-2.0.0.gem (100%)
Fetching: deckar01-task_list-2.0.1.gem (100%)
Compared versions: ["2.0.0", "2.0.1"]
  DIFFERENT files:
    2.0.0->2.0.1:
      * Changed:
            .travis.yml +0/-1
            app/assets/javascripts/task_list.coffee +1/-8
            bower.json +1/-1
            lib/task_list/version.rb +1/-1
            package.json +1/-1
            script/bootstrap +1/-1
            task_list.gemspec +2/-2
            test/unit/test_updates.coffee +10/-10
```

## Proposal

Add an option that outputs the line diff for the unpacked gem directories with syntax highlighting. There is probably a better UX that can be achieved by integrating the diff into the existing report format, but I plan to just call out of `git diff` when I try this on my fork.

```diff
$ gem compare --diff deckar01-task_list 2.0.0 2.0.1
Fetching: deckar01-task_list-2.0.0.gem (100%)
Fetching: deckar01-task_list-2.0.1.gem (100%)
Compared versions: ["2.0.0", "2.0.1"]

diff --git a/deckar01-task_list-2.0.0/.travis.yml b/deckar01-task_list-2.0.1/.travis.yml
index 4996b15..0cf6108 100644
--- a/deckar01-task_list-2.0.0/.travis.yml
+++ b/deckar01-task_list-2.0.1/.travis.yml
@@ -8 +7,0 @@ rvm:
-  - 1.9.3
diff --git a/deckar01-task_list-2.0.0/app/assets/javascripts/task_list.coffee b/deckar01-task_list-2.0.1/app/assets/javascripts/task_list.coffee
index 3fad38b..9d6d1f6 100644
--- a/deckar01-task_list-2.0.0/app/assets/javascripts/task_list.coffee
+++ b/deckar01-task_list-2.0.1/app/assets/javascripts/task_list.coffee
@@ -193,8 +193 @@ class TaskList
-    \s+                     # is followed by whitespace
-    (?!
-      \(.*?\)               # is not part of a [foo](url) link
-    )
-    (?=                     # and is followed by zero or more links
-      (?:\[.*?\]\s*(?:\[.*?\]|\(.*?\))\s*)*
-      (?:[^\[]|$)           # and either a non-link or the end of the string
-    )
+    \s                      # is followed by whitespace
diff --git a/deckar01-task_list-2.0.0/bower.json b/deckar01-task_list-2.0.1/bower.json
index cf77b36..ae800e6 100644
--- a/deckar01-task_list-2.0.0/bower.json
+++ b/deckar01-task_list-2.0.1/bower.json
@@ -3 +3 @@
-  "version": "2.0.0",
+  "version": "2.0.1",
diff --git a/deckar01-task_list-2.0.0/lib/task_list/version.rb b/deckar01-task_list-2.0.1/lib/task_list/version.rb
index 87a969f..3f6f80a 100644
--- a/deckar01-task_list-2.0.0/lib/task_list/version.rb
+++ b/deckar01-task_list-2.0.1/lib/task_list/version.rb
@@ -2 +2 @@ class TaskList
-  VERSION = [2, 0, 0].join('.')
+  VERSION = [2, 0, 1].join('.')
...
```

**Comment by deckar01** [comment]
The line diffs were readily available from Diffy, so I went ahead and hacked together a proof of concept.

<img width="759" alt="screen shot 2019-01-15 at 11 47 50 am" src="https://user-images.githubusercontent.com/3108007/51199052-7147e980-18bb-11e9-925f-dced1fdf82e6.png">

I copied the file list comparator and showed the line diffs under the file names in the report. After I got it working I realized that using a flag to modify the existing file comparator output would have been much more elegant. I don't full understand the other params that use that comparator, so I probably won't be able to improve this to PR quality myself.

https://github.com/fedora-ruby/gem-compare/compare/master...deckar01:9a59eda

**Comment by mattbrictson** [comment]
I'd love to see this 👍

**Comment by strzibny** [comment]
@deckar01 Looks nice, can you submit a PR? I review it.

**Comment by josephfrazier** [comment]
I'd also like to see this, so I took the liberty of opening a PR with @deckar01's changes from https://github.com/fedora-ruby/gem-compare/issues/17#issuecomment-454489378: https://github.com/fedora-ruby/gem-compare/pull/20

**Comment by strzibny** [comment]
Perfect, I'll have a look soon. Sorry for the wait!

**Comment by strzibny** [comment]
Merged as #21
